### ISSUE TITLE
Integrate new JsonCpp API instead of deprecated one

### ISSUE THREAD

**Issue by argenet** [scylladb/scylla#2989]
In JsonCpp 1.8.3, a bunch of classes have been deprecated in favour of newer/better ones:
https://github.com/open-source-parsers/jsoncpp/commit/ef16a35328217fc11e4595b257ec3c13bf66f55b 

This results in tons of warnings generated when building ScyllaDB:

```
In file included from ./caching_options.hh:26:0,
                 from ./schema.hh:39,
                 from ./mutation_partition.hh:33,
                 from ./mutation.hh:26,
                 from ./tests/mutation_assertions.hh:24,
                 from tests/row_cache_test.cc:27:
./json.hh: In function ‘seastar::sstring seastar::json::to_json(const Map&)’:
./json.hh:37:22: warning: ‘FastWriter’ is deprecated: Use StreamWriterBuilder instead [-Wdeprecated-declarations]
     Json::FastWriter writer;
                      ^~~~~~
In file included from /usr/include/json/json.h:12:0,
                 from ./json.hh:26,
                 from ./caching_options.hh:26,
                 from ./schema.hh:39,
                 from ./mutation_partition.hh:33,
                 from ./mutation.hh:26,
                 from ./tests/mutation_assertions.hh:24,
                 from tests/row_cache_test.cc:27:
/usr/include/json/writer.h:159:70: note: declared here
 class JSONCPP_DEPRECATED("Use StreamWriterBuilder instead") JSON_API FastWriter : public Writer {
                                                                      ^~~~~~~~~~
In file included from ./caching_options.hh:26:0,
                 from ./schema.hh:39,
                 from ./mutation_partition.hh:33,
                 from ./mutation.hh:26,
                 from ./tests/mutation_assertions.hh:24,
                 from tests/row_cache_test.cc:27:
./json.hh: In function ‘Map seastar::json::to_map(const sstring&, Map&&)’:
./json.hh:51:18: warning: ‘Reader’ is deprecated: Use CharReader and CharReaderBuilder instead [-Wdeprecated-declarations]
     Json::Reader reader;
                  ^~~~~~
In file included from /usr/include/json/json.h:11:0,
                 from ./json.hh:26,
                 from ./caching_options.hh:26,
                 from ./schema.hh:39,
                 from ./mutation_partition.hh:33,
                 from ./mutation.hh:26,
                 from ./tests/mutation_assertions.hh:24,
                 from tests/row_cache_test.cc:27:
/usr/include/json/reader.h:35:83: note: declared here
 class JSONCPP_DEPRECATED("Use CharReader and CharReaderBuilder instead") JSON_API Reader {
```

While this doesn't affect the build directly, it pollutes ninja output and makes it harder to read/find real compilation errors. Besides that, it is generally a good idea to go from the deprecated API towards the recommended one.

Reproducible with Fedora 27.

**Comment by avikivity** [comment]
Note that CharReader isn't available on Ubuntu 14.04, so we need dual code (or to backport newer jsoncpp to older Ubuntu)

**Comment by argenet** [comment]
We can #ifdef the code based on JsonCpp version available in json/version.h and get rid of the old code once we cease support for 14.04.
